ADC_irq/Core/Src/main.c

```diff
@@ -8,101 +8,129 @@
+	ADC1->CR 		|=	ADC_CR_ADVREGEN;		//enables the voltage regulator
+	ADC1->CFGR2 	|= 	ADC_CFGR2_CKMODE_0;		//sets ADC clock as PCLK/2
+	ADC->CCR 		|= 	ADC_CCR_LFMEN;			//low frequency mode enable
+	ADC1->SMPR 		|= 	ADC_SMPR_SMPR_1	|		//sets sample time for 12.5 ADC clock cycles
+						ADC_SMPR_SMPR_0	;		//because we need 10.5 ADC clock cycles at least
+	ADC1->CFGR1		|=	ADC_CFGR1_WAIT;			//enables wait mode to prevent overrun
+
+	ADC1->CR |= ADC_CR_ADCAL;					//starting the calibration
+	while(ADC1->CR & ADC_CR_ADCAL);				//we have to wait until ADCAL = 0 (Can be handled by interrupt)
+
+	ADC1->IER		|=	ADC_IER_EOCIE;			//enables End Of Conversion interrupt
+
+	//Enables IRQ and IRQ priority
+	NVIC_EnableIRQ(ADC1_COMP_IRQn);
+	NVIC_SetPriority(ADC1_COMP_IRQn, 0);
+
+	ADC1->ISR|= ADC_ISR_ADRDY;					//clear the ADRDY bit by programming it to 1
+	ADC1->CR |= ADC_CR_ADEN;					//then we enable the ADC
+	while(!(ADC1->ISR & ADC_ISR_ADRDY));		//and wait for it to be ready (Can be handled by interrupt)
+
+	/*												*
+	 * --- ADC CHANNEL CONFIG FOR V_REF READING --- *
+	 *												*/
 	if(ADC1->CR & ADC_CR_ADSTART)			//we have to be sure there's no ongoing conversion
 	{
 		ADC1->CR |= ADC_CR_ADSTP;			//if so, we have to set ADSTP bit
 		while(ADC1->CR & ADC_CR_ADSTP);		//and wait for it to clear
 	}
 	ADC1->CR |= ADC_CR_ADDIS;				//disables the ADC
 	while(ADC1->CR & ADC_CR_ADEN);
-	DISABLE_ALL_CHANNELS();					//and disables all channels
 
-	switch (Channel)
-	{
-		case CH_INT_TEMP:
-				ADC1->CHSELR |= ADC_CHSELR_CHSEL18;		//selecting the TSEN channel
-				ADC->CCR |= ADC_CCR_TSEN; 				//enables temperature sensor
-				Wait(TIME_10uSEC);						//we have to wait for the proper time for the Tsense to wake up
-		break;
-		case CH_V_REF:
-				ADC1->CHSELR |= ADC_CHSELR_CHSEL17;		//selecting the VREF channel
-				ADC->CCR |= ADC_CCR_VREFEN; 			//enables internal reference voltage
-		break;
-		default:
-				return (ADC_ERROR);	//ADC error
-		break;
-	}
+	ADC1->CHSELR = ADC_CHSELR_CHSEL17;		//selecting the VREF channel
+	ADC->CCR |= ADC_CCR_VREFEN; 			//enables internal reference voltage
 
-	ADC1->ISR|= ADC_ISR_ADRDY;				//clear the ADRDY bit by programming it to 1
 	ADC1->CR |= ADC_CR_ADEN;				//then we enable the ADC
 	while(!(ADC1->ISR & ADC_ISR_ADRDY));	//and wait for it to be ready. (Can be handled by interrupt)
 
+	/*									*
+	 * --- STARTING THE ADC READING --- *
+	 *									*/
+	ADC1->CR |= ADC_CR_ADSTART;				//starts the ADC
+
+	while(1)
+	{
+		if(flag_EOC)									//if THERE WAS an ADC interrupt
+		{
+			flag_EOC = 0;
+
+			switch (ADC1->CHSELR)
+			{
+				case ADC_CHSELR_CHSEL17:				//Internal reference voltage
+						v_ref = ADC1->DR;				//storing the data read in v_ref
+
+						//configuring the next channel to be read
+						ADC1->CHSELR = ADC_CHSELR_CHSEL18;
+						ADC->CCR 	|= ADC_CCR_TSEN; 	//enables temperature sensor
+						wait(TIME_10uSEC);				//we have to wait for the proper time for the Tsense to wake up

```

Comment:
I know that it would be a draw back in matters of energy consumption, but we can enable TSEN in the same place that we enable VREF. The positive side is that we will not need to wait TSEN get ready in the main loop.

Actually, thinking further on how this code is going to run, as you use the operator `|=` to enable TSEN, the second time the that code enter this case, TSEN will be enabled already.

My suggestion: Enable TSEN in the code initialisation and only change the channel to convert inside the switch case